### PR TITLE
Add soda-syncer base code

### LINKED ISSUE #423: Add Consistent Snapshot solution for local PV
**Issue/Feature Description:**
Add an experimental feature for consistent snapshot solution for Container Storages

**Why this issue to fixed / feature is needed(give scenarios or use cases):**
This feature is provided to give a DR solution to CSI

### PR DESCRIPTION
**Pull request by asifdxtreme**
**What type of PR is this?**
 /kind new feature

**What this PR does / why we need it**:
This PR adds the experimental `consistent snapshot` solution code to NBP project
**Which issue(s) this PR fixes**:

Fixes #423

**Test Report Added?**:
 /kind TESTED

**Special notes for your reviewer**:
This is an experimental feature so only positive flow has been tested, corner cases and specific coding guidelines will be addressed later

### CODE REVIEW

File: soda-syncer/Readme.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+ - [ ] CSI Meta-data sync between Soda and K8s
+
+
+*Note*: This is an experimental project and is not ready to be used in production.
+
+### Consistent Snapshot Solution for CSI Plugins
+Consistent Snapshot is considered as snapshots which are taken at regular intervals and pushed to cloud for DR Solutions.   
+Consistent snapshot is provided by few CSI Plugin Drivers, but most of them lack this feature, Soda provides this solution for CSI Plugin Drivers which gives local PV. The local PV is backed up by soda-syncer at regular interval as configured and backed up to the cloud of your choice.
+Soda leverages the Soda profile and [CSI Plug-N-Play](../csi-plug-n-play/) design to configure the snapshot policy and does the backup independently without platform support, currently this solution is available for K8s, and it doesn't require any operator/crd to add on this feature to existing CSI plugins.
+
```
Comment (kumarashit) [inline soda-syncer/Readme.md:16]: What is SODA proxy and how are we supporting here, needs clarity

File: soda-syncer/pkg/snapshot/worker.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	ticker := time.NewTicker(timeD* time.Second)
+	done := make(chan bool)
+	go func() {
+		for {
+			select {
+			case <-done:
+				return
+			case t := <-ticker.C:
+				fmt.Println("Backup Started at", t)
+				cmdToDoBackup := "restic backup "+mountPointToBeBackedUp
```
Comment (kumarashit) [inline soda-syncer/pkg/snapshot/worker.go:71]: We assume that restic is installed. Can we document it for the user as a pre-requisite? Do we need to include it in SODA installer instead of user to install restic?

File: soda-syncer/pkg/snapshot/worker.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+	"fmt"
+	"os"
+	"time"
+
+	"net/http"
+	"os/exec"
+)
+
+type SnapshotProfile struct {
+	AwsAccesskey string `json:"AWS_ACCESS_KEY_ID"`
```
Comment (kumarashit) [inline soda-syncer/pkg/snapshot/worker.go:28]: This can work for any cloud and not specific to AWS, right? if we provide the AK, SK and bucket URL (as restic repository) then it can work for any cloud.

Comment (kumarashit) [reply soda-syncer/pkg/snapshot/worker.go:28]: If yes, then we can name the key as SECRET_KEY and ACCESS_KEY

File: soda-syncer/pkg/snapshot/worker.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+	//TODO Configure grep based on the pvc name recieved in this request
+	cmdToGetPVCMountPoint := "df -h --output=target | grep csi"
+	cmdToGetFileSystemMountPoint, errno := exec.Command("bash", "-c", cmdToGetPVCMountPoint).Output()
+	if errno != nil {
+		 fmt.Sprintf("Failed to execute command: %s", cmdToGetFileSystemMountPoint)
+	}
+	mountPointToBeBackedUp := string(cmdToGetFileSystemMountPoint)
+	fmt.Println("The MountPoint to be backed up : ",mountPointToBeBackedUp )
+
+	fmt.Println("---------------------Backing Up ---------------------")
```
Comment (kumarashit) [inline soda-syncer/pkg/snapshot/worker.go:52]: Can these logs be redirected to file? this will keep spitting over STDOUT

File: soda-syncer/pkg/snapshot/worker.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+				cmdOutputForBackup, errBck := exec.Command("bash", "-c", cmdToDoBackup).Output()
+				if errBck != nil {
+					fmt.Sprintf("Failed to execute command: %s", string(cmdOutputForBackup))
+				}
+				fmt.Println("Backup Success : ", string(cmdOutputForBackup))
+
+				fmt.Println("---------------------SnapShots---------------------")
+				cmd := exec.Command("restic", "snapshots" )
+				cmd.Stderr = os.Stdout
+				cmd.Stdout = os.Stdout
```
Comment (kumarashit) [inline soda-syncer/pkg/snapshot/worker.go:81]: stderr and stdout can be captured into log file

File: soda-syncer/Readme.md (hunk 1 above)
Comment (asifdxtreme) [reply soda-syncer/Readme.md:16]: soda-proxy is an exsisting component and I have added the flow details and link to the module

File: soda-syncer/pkg/snapshot/worker.go (hunk 3 above)
Comment (asifdxtreme) [reply soda-syncer/pkg/snapshot/worker.go:28]: For each cloud the key is different, so we cannot generalize the key

File: soda-syncer/pkg/snapshot/worker.go (hunk 2 above)
Comment (asifdxtreme) [reply soda-syncer/pkg/snapshot/worker.go:71]: Added the quick start guide with these details

File: soda-syncer/pkg/snapshot/worker.go (hunk 4 above)
Comment (asifdxtreme) [reply soda-syncer/pkg/snapshot/worker.go:52]: If we start the soda-syncer using the script given in quickstart guide then all the logs will be re-directed to log file, later on we can follow the similar logging mechanism as soda in later phase

File: soda-syncer/pkg/snapshot/worker.go (hunk 5 above)
Comment (asifdxtreme) [reply soda-syncer/pkg/snapshot/worker.go:81]: If we start the soda-syncer using the script given in quickstart guide then all the logs will be re-directed to log file, later on we can follow the similar logging mechanism as soda in later phase

File: soda-syncer/pkg/snapshot/worker.go (hunk 3 above)
Comment (kumarashit) [reply soda-syncer/pkg/snapshot/worker.go:28]: SODA multicloud bucket management works for most of the public cloud with AK/SK/URI. We can keep the key name generic and the values can be based upon

Comment (asifdxtreme) [reply soda-syncer/pkg/snapshot/worker.go:28]: For reference : https://restic.readthedocs.io/en/latest/040_backup.html#environment-variables

File: soda-syncer/Readme.md (hunk 1 above)
Comment (kumarashit) [reply soda-syncer/Readme.md:16]: ok

File: soda-syncer/pkg/snapshot/worker.go (hunk 5 above)
Comment (kumarashit) [reply soda-syncer/pkg/snapshot/worker.go:81]: ok

File: soda-syncer/pkg/snapshot/worker.go (hunk 3 above)
Comment (kumarashit) [reply soda-syncer/pkg/snapshot/worker.go:28]: got it thanks

**Review by kumarashit — APPROVED** [review]
LGTM

**Comment by sushanthakumar** [comment]
LGTM

**Review by skdwriting — APPROVED** [review]
LGTM
